core/src/main/java/io/questdb/cairo/wal/seq/TableWriterPressureControlImpl.java

```diff
@@ -0,0 +1,164 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.wal.seq;
+
+import io.questdb.cairo.wal.TableWriterPressureControl;
+import io.questdb.std.Rnd;
+import io.questdb.std.datetime.millitime.MillisecondClock;
+
+// Implements TableWriterPressureControl interface and regulates number of transactions and partitions
+// TableWriter uses when applying wal transactions.
+// See TableWriterPressureControl interface for more information.
+public class TableWriterPressureControlImpl implements TableWriterPressureControl {
+    // To be used in multiple threads, safe and values will still be random
+    private static final Rnd MEM_PRESSURE_RND = new Rnd();
+    private static final int PARTITION_COUNT_SCALE_DOWN_FACTOR = 4;
+    private static final int PARTITION_COUNT_SCALE_UP_FACTOR = 4;
+    private static final int TXN_COUNT_SCALE_DOWN_FACTOR = 4;
+    private static final int TXN_COUNT_SCALE_UP_FACTOR = 1000;
+    private final int backOffWaiMs;
+    private final MillisecondClock millisecondClock;
+    private long inflightBlockRowCount;
+    private long inflightTxnCount;
+    private long maxBlockRowCount = Integer.MAX_VALUE;
+    private int maxRecordedInflightPartitions = 1;
+    // positive int: holds max parallelism
+    // negative int: holds backoff counter
+    private int memoryPressureRegulationValue = Integer.MAX_VALUE;
+    private long walBackoffUntilEpochMs = Long.MIN_VALUE;
+
+    public TableWriterPressureControlImpl(int backOffWaiMs, MillisecondClock millisecondClock) {
```

Comment:
Wai? Can we maintain "configuration" here to make the config reloadable without database bounce?

Reply:
Done in [ef5b832](https://github.com/questdb/questdb/pull/5347/commits/ef5b832b764a7c4e06e5042fc5139635f5e9bd34)